gateway/ln-gateway/src/gatewayd/actor.rs

```diff
@@ -147,7 +154,36 @@ impl GatewayActor {
 
                     let amount_msat = Amount::from_msats(outgoing_amount_msat);
 
-                    match actor.buy_preimage_internal(&hash, &amount_msat).await {
+                    let (outpoint, contract_id) =
+                        match actor.buy_preimage_internal(&hash, &amount_msat).await {
+                            Ok((outpoint, contract_id)) => (outpoint, contract_id),
+                            Err(e) => {
+                                error!("Failed to buy preimage: {:?}", e);
+                                // Note: this specific complete htlc requires no futher action.
+                                // If we fail to send the complete htlc message, or get an error
+                                // result, lightning node will still
+                                // cancel HTCL after expiry period lapses.
+                                // Result can be safely ignored.
+                                // TODO: make sure this succeeded?
+                                let _ = lnrpc_copy
+                                    .complete_htlc(CompleteHtlcsRequest {
+                                        intercepted_htlc_id,
+                                        action: Some(Action::Cancel(Cancel {
+                                            reason: e.to_string(),
+                                        })),
+                                    })
```

Comment:
hm. I don't think `buy_preimage_internal` should have anything to do with `complete_htlc!` cc @justinmoon

Reply:
fixing

Reply:
what do you mean by "fixing"?

Reply:
> Re: "fixing"

I was concerned by the application of `complete_htlc` rpc call to cancel htlc in case `buy_preimage_internal` failed. My intuition was, _`buy_preimage_internal` should not have anything to do with `complete_htlc` rpc_. So **fixing** meant trying to resolve the failed buy preimage by some other means, other than complete htlc

As a follow up, I figured we actually have to call `complete_htlc`, otherwise we'd have a long pending htlc intercept scenario, before the htlc lapses it's natural expiry and gets cancelled by the LN node.

Further follow up will involve removing internal payments as a feature of gateways. See #613